KTAS-5: Apply Raft for distributed tasks

dhuynh-Kazoo opened the pull request:
Apply a new way to distribute task jobs by using `ra` implementation. 
The following ticket: https://2600hz-commercial.atlassian.net/browse/KTAS-5.

---
Adding Karl's comment:

> my intention with regards to Raft was to use it to elect a worker when there are multiple.  IE: on start up of each task use raft to determine if there is another worker already for that task and if stop/defer to it.  otherwise take leadership of that task via raft and execute
> [10:31](https://2600hz.slack.com/archives/D0535CAVDAT/p1693323113967149)
> you could take a look at how fax workers do the job distribution using raft to "lock" each account to a node when they pickup a new job
> [10:33](https://2600hz.slack.com/archives/D0535CAVDAT/p1693323223343529)
> for tasks they aren't necessarily as associated with accounts as faxes so we might need to lock at a different level, such as the task type, without looking into details I believe the best would be to acquire a raft leadership as a combination of task type and trigger (IE: task X running every Y) releasing each time that specific instance completes
> [10:35](https://2600hz.slack.com/archives/D0535CAVDAT/p1693323307316319)
> I would NOT want tasks to start up and get a raft lock for that instance, blocking all other tasks instances, as that would put every periodic task on one node and not be as robust as negotiating each trigger.
> 

---
update 04/23/2024.
[DRAFT] Summary about the implementation and question would need your answer/clarification:
- How do we start Erlang nodes in the cluster and how do they connect to each other to become a cluster?

- Summary for the idea(I'll update the detailed later). But it would be great if you can help me the review the idea. Thank you!! So Have all nodes connected then they can start an election, and then we have leader and followers. Most of the tasks are periodic type with gener_server behavior. So my idea is to send a start command for each type of task to the ra leader,  the leader starts the task on the random node and updates the task on it state _#{task_type => Type, node => node, pid => pid()}_. The leader will monitor the task and start it if crashes (choose a random node to start). 
Duplicated requests can be checked by ra leader by lookup the state with task type and pid on a node, if a duplicate is found, just rejects the command. Otherwise, implement that command

icehess committed:
tinkering with rebar3 as alternative experimental build system (#36)

add rebar.config

icehess committed:
bump orb to use custom couchdb image (#35)

danielfinke committed:
gitignore `erlang_ls.config` per 2600hz/kazoo5@56c4cad (#33)

icehess committed:
update orb to consider os-release in cache keys (#37)

With upcoming changes, we need another variable to invalidate caches.

This adds /etc/os-release, uname -pmo and others together and uses them
as cache key in CI.

dhuynh-Kazoo commented:
hi @icehess, could you take the time to look at this PR?

I moved all sub-task modules to `task_ra_sup` and have only 4 functions for interaction with RA cluster: `start, stop, remove and sync_state`(for kz_tasks_scheduler because I noticed that only the scheduler needs to update to worker's state). So the ra server status is updated by the `tick/2`(spawn process every interval time to get `kz_tasks_scheduler` state ).

The other background tasks(some are time triggered and some are time interval set up) are independent and are not linked to `kz_task_scheduler`, so I just moved the worker processes from `task_sup` to my new `task_ra_sup`.

jamesaimonetti commented on src/kz_tasks_scheduler.erl:
```diff
@@ -16,8 +16,7 @@
         ,stop/1
-        ,remove/1
-        ]).
+        ,remove/1]).
```
revert this, project standards have these end-of-list  brackets dropped to newline

jamesaimonetti commented on src/kz_account_crawler.erl:
```diff
@@ -14,7 +14,8 @@
 -export([start_link/0]).
--export([stop/0, check/1]).
+-export([stop/0, check/1
+        ,crawl_account/1]).
```
drop end-of-list to newline

jamesaimonetti commented on src/kz_account_crawler.erl:
```diff
@@ -128,7 +129,7 @@ handle_info({'timeout', Ref, _Msg}, #state{cleanup_ref = Ref
                                           ,account_ids = [AccountId]
                                           }=State) ->
-    _ = crawl_account(AccountId),
+    _ = tasks_ra_action:pipeline_command({crawl_account, AccountId}),
```
ticky atoms

jamesaimonetti commented on src/kz_notify_resend.erl:
```diff
@@ -199,7 +199,8 @@ handle_info({'timeout', Ref, _Msg}, #state{timer_ref = Ref}=State) ->
         {'ok', Pendings} ->
             lager:info("processing ~b pending notifications", [length(Pendings)]),
-            _ = kz_process:spawn(fun () -> process_then_next_cycle([kz_json:get_value(<<"doc">>, J) || J <- Pendings]) end),
+            ProcessingFun = fun () -> process_then_next_cycle([kz_json:get_value(<<"doc">>, J) || J <- Pendings]) end,
```
Can you change this to `kz_json:get_json_value` to use the typed accessor

jamesaimonetti commented on src/kz_tasks_scheduler.erl:
```diff
@@ -570,7 +569,9 @@ handle_call_start_task(Task=#{id := TaskId
-    try kz_process:spawn_link(fun Worker:start/3, [TaskId, API, ExtraArgs]) of
+    try tasks_ra_action:process_command({'spawn_link'
+                                        ,fun Worker:start/3
+                                        ,[TaskId, API, ExtraArgs]}) of
```
drop `}` to newline, then `]` to a new line, then `of` to a line under that
```
    try tasks_ra_action:process_command({'spawn_link'
                                        ,fun Worker:start/3
                                        ,[TaskId, API, ExtraArgs]
                                        }
                                       )
    of
```

jamesaimonetti commented on src/tasks_ra_action.erl:
```diff
@@ -0,0 +1,74 @@
+%%%-------------------------------------------------------------------
+%%% @author dhuynh
+%%% @copyright (C) 2023, <COMPANY>
```
Fix header comment to match project standards

jamesaimonetti commented on src/tasks_ra_action.erl:
```diff
@@ -0,0 +1,74 @@
+%%% Created : 24. Jul 2023 7:23 AM
+%%%-------------------------------------------------------------------
+-module(tasks_ra_action).
+-author("dhuynh").
```
remove

jamesaimonetti commented on src/tasks_ra_action.erl:
```diff
@@ -0,0 +1,74 @@
+%% API
+-export([start/1
+        ,stop/1
+        ,remove/1]).
```
drop end-of-list to newline

jamesaimonetti commented on src/tasks_ra_node.erl:
```diff
@@ -0,0 +1,168 @@
+-spec rand_node([node()]) -> node().
+rand_node([]) -> node();
+rand_node(Nodes) ->
+    Size = length(Nodes),
```
You might take the list of nodes and use kz_term:shuffle_list(Nodes) and hd() that

jamesaimonetti reviewed (changes requested):
Some first checks

dhuynh-Kazoo commented:
Hi James, @jamesaimonetti changes with code format are updated

dhuynh-Kazoo commented:
Hi @jamesaimonetti and @icehess, could you help me with this concern? 
Because we have some workers with gen_server behavior in the tasks app, if we replicate them in all `ra` servers, it would cause duplication. So should we have a client node manages the requests and send them the `ra leader`, then `ra leader `will do the replication logic when ra servers execute?

dhuynh-Kazoo commented:
Hi @icehess and @jamesaimonetti, could you please review this pull request? I have made some changes for the ra_machine, workers are distributed in random node. When a function is called in a node, it will check if the server resides in current node, the execute it. Otherwise, check other nodes and use rpc call to execute it

jamesaimonetti commented on src/kt_number_crawler.erl:
```diff
@@ -64,7 +64,14 @@ start_link() ->
 -spec stop() -> 'ok'.
 stop() ->
-    gen_server:cast(?SERVER, 'stop').
+    case whereis(?MODULE) of
```
I don't understand this check. start_link + init suggest the server always starts, and tasks_sup has it as a child with no conditions.

Not sure why we'd want to stop remote crawlers too? Perhaps instead a `tasks_maintenance` command to stop the local server and one to stop remote servers too?

But if you can explain when we'd want to stop the server, that'd be helpful too.

dhuynh-Kazoo replied:
Hi @jamesaimonetti, I have checked the task maintenance modules and other places so we do not use the `stop` functions. And they were added at the beginning when files were added. 

I think we can remove them since the task_sup manages the workers not using `stop`

jamesaimonetti commented on src/kt_port_request_crawler.erl:
```diff
@@ -50,7 +52,14 @@ start_link() ->
 -spec stop() -> 'ok'.
 stop() ->
-    gen_server:cast(?MODULE, 'stop').
+    case whereis(?MODULE) of
```
Same questions here as for kt_number_crawler

jamesaimonetti commented on src/kz_account_crawler.erl:
```diff
@@ -52,20 +53,38 @@ start_link() ->
-    end;
 check(Account) ->
-    check(kz_term:to_binary(Account)).
+    BinaryAccount =
```
I actually prefer to use function clauses for these. I think
```erlang
check(<<Account/binary>>) -> ...;
check(Account) ->
    check(kz_term:to_binary(Account)).
````

is more clear (as you know the first clause is working with a binary() and subsequent clauses do data transformation before the happy path clause

jamesaimonetti commented on src/kz_account_crawler.erl:
```diff
@@ -52,20 +53,38 @@ start_link() ->
+                                                       ,[BinaryAccount]);
+        _Pid ->
+            AccountId = kzs_util:format_account_id(BinaryAccount),
+            case kz_datamgr:open_doc(?KZ_ACCOUNTS_DB, AccountId) of
```
Use `kzd_accounts:fetch/1`. The account doc in the account database is the authoritative doc; the account doc in the `accounts` database is not.

jamesaimonetti commented on src/kz_notify_resend.erl:
```diff
@@ -134,21 +148,35 @@ running(Pid) ->
+                                                       ,?FUNCTION_NAME
+                                                       ,[Id]);
+        _ ->
+            case kz_datamgr:open_doc(?KZ_PENDING_NOTIFY_DB, Id) of
```
`kz_datamgr:open_cache_doc/2` should be used unless you have a good reason not to

jamesaimonetti commented on src/kz_notify_resend.erl:
```diff
@@ -134,21 +148,35 @@ running(Pid) ->
+        _ ->
+            case kz_datamgr:open_doc(?KZ_PENDING_NOTIFY_DB, Id) of
+                {'ok', JObj} ->
+                    kz_log:put_callid(kz_json:get_value(<<"payload">>, JObj)),
```
What is the expected value of `payload`? Use a kz_json typed accessor to match

dhuynh-Kazoo replied:
Hi Jame, I do not know the value of the payload because I do not have the case to test it in my local. Could you please give me an example of this? 

For ` Use a kz_json typed accessor to match` do you mean using a specific key?

jamesaimonetti commented on src/tasks.hrl:
```diff
@@ -12,8 +12,25 @@
-
 -define(OUTPUT_CSV_HEADER_ERROR, <<"error">>).
 
+-define(START_WORKER_RPC_CALL_TIMEOUT, 10 * ?MICROSECONDS_IN_SECOND).
```
Generally I like to see the time unit appended to the var name `START_WORKER_RPC_CALL_TIMEOUT_US` for microseconds here makes sense I think?

jamesaimonetti commented on src/tasks_ra_node.erl:
```diff
@@ -0,0 +1,181 @@
+
+-include("tasks.hrl").
+
+-define(TICK_INTERVAL, 20 * ?MILLISECONDS_IN_SECOND).
```
`TICK_INTERVAL_MS`

jamesaimonetti commented on src/tasks_sup.erl:
```diff
@@ -67,4 +77,4 @@ init([]) ->
     SupFlags = {RestartStrategy, MaxRestarts, MaxSecondsBetweenRestarts},
 
-    {'ok', {SupFlags, ?CHILDREN}}.
+    {'ok', {SupFlags, []}}.
```
Oh, I see. You should probably remove the CHILDREN macro. The whereis checks make more sense now.

jamesaimonetti reviewed (changes requested):
Some quick checks

dhuynh-Kazoo commented on src/tasks_ra_node.erl:
```diff
@@ -0,0 +1,181 @@
+-spec join() -> 'ok' | {'error', any()}.
+join() ->
+    RaConf = make_ra_conf(id()),
+    case ra:start_server(RaConf) of
```
Hi @jamesaimonetti, these tasks_ra_* modules are replicated from `fax` app.

The idea of raft is that multiple nodes are connected and then vote for an election. I see an issue that may be a bug with the current implementation. When the `tasks` app starts, the code checks for connected nodes if no connected node, it becomes the leader. 

This is the race condition for the case that we have multiple erlang nodes, every node try to become the leader as they still not know others up and connected.

I think we have prefined cluster nodes in the configuration, when tasks app starts, it tries to connect each other with a timeout. If none of these nodes are found, it will become a leader. The cluster node will be set in the system_config db, please let me know your idea.

dhuynh-Kazoo replied:
@icehess, please help to check and maybe share the way we used `ra` between erlang nodes in a cluster

danielfinke replied:
As I understand it, this is an inherent limitation of the Raft protocol or perhaps `ra`'s implementation of it. It is required that every node knows all cluster members ahead of time or at least that every node knows of a seed node that is started first: https://github.com/rabbitmq/ra/blob/main/docs/internals/INTERNALS.md#cluster-setup

jamesaimonetti committed:
Remove the stacktrace macro

Commit:
update shipyard config to support rockylinux

Commit:
enable circle-ci support for rocky linux

jamesaimonetti committed:
Update python scripts

icehess committed:
rebar3: run scripts pre-compile only if headers are missing (#40)

danielfinke commented on src/tasks_ra_node.erl:
For `maybe_join`, `join`, `join_add_member`, and `start_new` take a look at what I have in this diff: https://github.com/2600hz/kazoo-qubicle/commit/155f9e6b448d3815bc6c5a76fdc43318da32bf93#diff-7c934d4272b6f4cb45840b352d92da59f1c969271593a1d918041e83eb345845. You'll see I have added a couple extra checks and I'm leaving lots of room for `badmatch` errors (intentionally). The idea is that if _any_ of these checks fail, the app should fail to start. Try those things out here.

Since Raft is strict about knowing the nodes ahead of time, the necessary deployment strategy for Ra apps is to make sure the Erlang nodes are connected to each other before the first start of the app. The cluster then forms successfully. For subsequent restarts, because the existing Ra server is restarted first, there is no need for the Erlang nodes to be connected. The Ra server will work fine once they are.

Decommissioning of a Ra app requires an explicit leave of the cluster (see my `leave_and_force_delete_server`).

dhuynh-Kazoo committed:
Using raft for distributed tasks with locking

dhuynh-Kazoo committed:
Apply raft lock for periodic task workers

manushiperera committed:
KZOO-550: Initiate MODB rollover for manual presence IDs (#41)

Previously, manual presence records were stored in a single document
per master account.

Now, the system has been improved to track manual presence in
individual documents for each presence ID, managed within the "modb"
database.

At the end of each month, the latest manual presence documents for
each presence ID are rolled over to the next month's "modb" through
the newly defined presence rollover task.

dhuynh-Kazoo committed:
KTAS-7: Handle importing numbers from CSV (#47)

Put E164 in list as kt_task_worker:loop/2 fetch csv row by row

dhuynh-Kazoo committed:
KPIV-7: fix calls to knm_ops require a list of numbers (#48)

romanat-ooma committed:
bump copyright
